packages/smooth_app/lib/query/products_preload_helper.dart

```diff
@@ -1,23 +1,26 @@
 import 'dart:io';
 import 'package:openfoodfacts/openfoodfacts.dart';
 import 'package:smooth_app/database/dao_product.dart';
-import 'package:smooth_app/database/local_database.dart';
 import 'package:smooth_app/query/product_query.dart';
 
 class PreloadDataHelper {
-  PreloadDataHelper(this.localDatabase);
-  LocalDatabase localDatabase;
-
-  Future<String> preloadData() async {
-    final DaoProduct daoProduct = DaoProduct(localDatabase);
+  PreloadDataHelper(this.daoProduct);
+  DaoProduct daoProduct;
+  Future<String> getTopProducts() async {
+    List<String> allProductCodes = await daoProduct.getAllKeys();
+    final Map<String, Product> allProducts =
+        await daoProduct.getAll(allProductCodes);
+    allProducts.removeWhere(
+        (String key, Product value) => value.knowledgePanels == null);
+    allProductCodes = allProducts.keys.toList();
     final List<ProductField> fields = ProductQuery.fields;
     fields.remove(ProductField.KNOWLEDGE_PANELS);
```

Comment:
Actually that's not OK: you're also removing an item from the main `fields`.
You should copy all the fields you need instead, into your own list.

Reply:
Had done a `for each` iteration but as a left suggestion I changed it 

Might be something like this would work fine 
``` dart  
  final List<String> allProductCodes = await daoProduct.getAllKeys();
    final Map<String, Product> allProducts =
        await daoProduct.getAll(allProductCodes);
    allProducts.forEach((String code, Product product) {
      if (product.knowledgePanels != null) {
        allProductCodes.remove(code);
      }
    });
```

Reply:
You haven't implemented a fix, have you?